sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1148,9 +1148,9 @@ object CollapseWindow extends Rule[LogicalPlan] {
  */
 object TransposeWindow extends Rule[LogicalPlan] {
   private def compatiblePartitions(ps1 : Seq[Expression], ps2: Seq[Expression]): Boolean = {
-    ps1.length < ps2.length && ps2.take(ps1.length).permutations.exists(ps1.zip(_).forall {
```

Comment:
I'd consider this as a perf bug...

Reply:
You mean we should only optimize the performance without changing the logic？Maybe we could find a way to avoid the permutation

Reply:
I mean the previous code is kind of buggy. This PR is more like a bug fix instead of perf improvement and we should backport it.